[TableGen] Extend direct lookup to instruction values in generic tables.

nvjle opened the pull request:
Currently, for some tables involving a single primary key field which is integral and densely numbered, a direct lookup is generated rather than a binary search. This patch extends the direct lookup function generation to instructions, where the integral value corresponds to the instruction's enum value.

While this isn't as common as for other tables, it does occur in at least one downstream backend and one in-tree backend.

Added a unit test and minimally updated the documentation.

nvjle committed:
[TableGen] Extend direct lookup to instruction values in generic tables.

Currently, for some tables involving a single primary key field which is
integral and densely numbered, a direct lookup is generated rather than
a binary search. This patch extends the direct lookup function generation
to instructions, where the integral value corresponds to the instruction's
enum value.

While this isn't as common as for other tables, it does occur in at least
one downstream backend and one in-tree backend.

Added a unit test and minimally updated the documentation.

wangpc-pp commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -646,11 +671,23 @@ void SearchableTableEmitter::collectTableEntries(
+  if (SawInstructionField && InstrEnumValueMap.empty()) {
+    CodeGenTarget Target(Records);
+    unsigned Num = 0;
+    for (const CodeGenInstruction *Inst : Target.getInstructionsByEnumValue())
```
I think we should add a `CodeGenTarget::getInstrToIntMap` (or other names?) to get this instr->int mapping, so that we can reuse these logics and sync with other emittters (like you said, `InstrInfoEmitter::emitEnums`).

nvjle replied:
As you suggested, I have moved this map into `CodeGenTarget::InstrToIntMap` and added a lookup method `CodeGenTarget::getInstrIntValue`. The map is now populated in `CodeGenTarget::computeInstrsByEnum` right after the sort that creates the order in the first place. 

Overall this is cleaner and simplified the patch as well. Thanks again for the suggestions.

In a follow-up patch, as you alluded to, I could use the new query in `InstrInfoEmitter::emitEnums` (and elsewhere if applicable) to reuse/sync the logic.

wangpc-pp commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -638,6 +662,7 @@ void SearchableTableEmitter::collectTableEntries(
 
   Record *IntrinsicClass = Records.getClass("Intrinsic");
   Record *InstructionClass = Records.getClass("Instruction");
+  bool SawInstructionField = false;
```
SawInstructionField->HasInstructionField?

nvjle replied:
I have removed this variable since it is no longer necessary due to having moved the map population logic.

wangpc-pp commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -207,12 +213,17 @@ class SearchableTableEmitter {
 // known, return that numeric value.
-static int64_t getNumericKey(const SearchIndex &Index, Record *Rec) {
+static int64_t getNumericKey(const SearchIndex &Index, Record *Rec,
+                             InstrEnumMapT &InstrEnumMap) {
```
A style nit(just mine). What about making `getNumericKey` a class function like `getIntrinsic`, so that we don't need to pass `InstrEnumMap`?

nvjle replied:
Done. Although I removed the `InstrEnumMap` parameter, it is still useful to do this since we need to invoke the new query function `Target->getInstrIntValue`.

wangpc-pp reviewed (commented):
I think this is a great idea generally!

nvjle committed:
Address review comments.

Move the instruction-to-int map into CodeGenTarget::InstrToIntMap and
add CodeGenTarget::getInstrIntValue query method.

Populate the map in CodeGenTarget::ComputeInstrsByEnum just below
where the order is created in the first place.

nvjle commented:
> I think this is a great idea generally!

Thank you for the review comments, it is appreciated. I've made changes according to your suggestions below.

wangpc-pp commented on llvm/utils/TableGen/SearchableTableEmitter.cpp:
```diff
@@ -663,6 +682,8 @@ void SearchableTableEmitter::run(raw_ostream &OS) {
   // Emit tables in a deterministic order to avoid needless rebuilds.
   SmallVector<std::unique_ptr<GenericTable>, 4> Tables;
   DenseMap<Record *, GenericTable *> TableMap;
+  if (Records.getClass("Instruction"))
```
```suggestion
  if (!Records.getAllDerivedDefinitions("Instruction").empty())
```
Do this only if there are `Instruction` definitions instead of `Instruction` class definition.

nvjle replied:
Done.
I've added the derived definitions guard as suggested. I needed to leave in the original guard since the derived definitions lookup will fail with "The class 'Instruction' is not defined" for clients that do not happen to have a Instruction class at all.

If I could also trouble you (or anyone else) to merge on my behalf, I'd appreciate that.

wangpc-pp replied:
Oh sorry, I forgot it, we can use `getAllDerivedDefinitionsIfDefined` here.
> If I could also trouble you (or anyone else) to merge on my behalf, I'd appreciate that.

Yeah, no problem!

nvjle replied:
Done (used `getAllDerivedDefinitionsIfDefined`).

wangpc-pp reviewed (approved):
LGTM with nits.

nvjle committed:
Address review nit.
Add derived definitions guard.

nvjle committed:
Address review comments.

Use getAllDerivedDefinitionsIfDefined.

topperc commented on llvm/utils/TableGen/CodeGenTarget.cpp:
```diff
@@ -538,6 +538,13 @@ void CodeGenTarget::ComputeInstrsByEnum() const {
+  // InstrInfoEmitter::emitEnums.
+  assert(InstrToIntMap.empty());
+  unsigned Num = 0;
+  for (const CodeGenInstruction *Inst : InstrsByEnum)
```
Sorry I'm late to this, but why not make this a field of CodeGenInstruction? We already have a map from Record to CodeGenInstruction.

nvjle replied:
> Sorry I'm late to this, but why not make this a field of CodeGenInstruction? We already have a map from Record to CodeGenInstruction.

Thanks for your comments @topperc, much appreciated. I'll make your suggested change in a new PR shortly, with you as reviewer/committer if you wouldn't mind.

piotrAMD commented:
Looks the code assumes that `Index.Fields[0].Enum` and `Index.Fields[0].IsInstruction` are exclusive. 

This is not universally true. For example `WMMAOpcode2AddrMappingTable` in AMDGPU target has both of them set.

This can cause problems, because `SearchableTableEmitter::compareBy()` or `primaryRepresentation()` check `IsInstruction` first and `Enum` next, but `SearchableTableEmitter::getNumericKey()` checks `Enum` first and `IsInstruction` next.

An immediate fix for that issue would be to switch the conditions around in `getNumericKey` so it checks `Instruction` before checking `Enum` maintainig consistency. Not sure how future-proof that would be, though.

piotrAMD commented:
CC @nvjle (not sure if notifications are sent without explicit mention)

nvjle commented:
> CC @nvjle (not sure if notifications are sent without explicit mention)

Thanks @piotrAMD, I'll look into this in detail sometime this week. At first glance it does seem there may be a potential problem, but there is no actual failure that I can see. All testing is good thus far, including for AMDGPU. But we surely want this to be bullet proof.

piotrAMD commented:
> > CC @nvjle (not sure if notifications are sent without explicit mention)
> 
> Thanks @piotrAMD, I'll look into this in detail sometime this week. At first glance it does seem there may be a potential problem, but there is no actual failure that I can see. All testing is good thus far, including for AMDGPU. But we surely want this to be bullet proof.

Thanks, I'll try to come up with a reproducer.

piotrAMD commented:
Thanks! I tried, but wasn't able to come up with a reproducer.
